wire-ios-data-model/Source/MLS/CommitSender.swift

```diff
@@ -0,0 +1,242 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import Combine
+import WireCoreCrypto
+
+// sourcery: AutoMockable
+public protocol CommitSending {
+
+    func sendCommitBundle(
+        _ bundle: CommitBundle,
+        for groupID: MLSGroupID
+    ) async throws -> [ZMUpdateEvent]
+
+    func sendExternalCommitBundle(
+        _ bundle: CommitBundle,
+        for groupID: MLSGroupID
+    ) async throws -> [ZMUpdateEvent]
+
+    func onEpochChanged() -> AnyPublisher<MLSGroupID, Never>
+
+}
+
+public actor CommitSender: CommitSending {
+
+    // MARK: - Properties
+
+    private let coreCryptoProvider: CoreCryptoProviderProtocol
+    private let notificationContext: NotificationContext
+    private let actionsProvider: MLSActionsProviderProtocol
+    private let onEpochChangedSubject = PassthroughSubject<MLSGroupID, Never>()
+
+    private var coreCrypto: SafeCoreCryptoProtocol {
+        get throws {
+            try coreCryptoProvider.coreCrypto(requireMLS: true)
+        }
+    }
+
+    // MARK: - Life cycle
+
+    public init(
+        coreCryptoProvider: CoreCryptoProviderProtocol,
+        notificationContext: NotificationContext
+    ) {
+        self.init(
+            coreCryptoProvider: coreCryptoProvider,
+            notificationContext: notificationContext,
+            actionsProvider: nil
+        )
+    }
+
+    init(
+        coreCryptoProvider: CoreCryptoProviderProtocol,
+        notificationContext: NotificationContext,
+        actionsProvider: MLSActionsProviderProtocol? = nil
+    ) {
+        self.coreCryptoProvider = coreCryptoProvider
+        self.notificationContext = notificationContext
+        self.actionsProvider = actionsProvider ?? MLSActionsProvider()
+    }
+
+    // MARK: - Public interface
+
+    public func sendCommitBundle(
+        _ bundle: CommitBundle,
+        for groupID: MLSGroupID
+    ) async throws -> [ZMUpdateEvent] {
+
+        do {
+            WireLogger.mls.info("sending commit bundle for group (\(groupID.safeForLoggingDescription))")
+            let events = try await sendCommitBundle(bundle)
+
+            WireLogger.mls.info("merging commit for group (\(groupID.safeForLoggingDescription))")
+            try await mergeCommit(in: groupID)
+
+            return events
+
+        } catch let error as SendCommitBundleAction.Failure {
+            WireLogger.mls.warn("failed to send commit bundle: \(String(describing: error))")
+
+            let recoveryStrategy = CommitError.Recovery(from: error)
```

Comment:
Should `CommitError.Recovery` also be named `CommitError.RecoveryStrategy` instead?

Reply:
done https://github.com/wireapp/wire-ios/pull/787/commits/b68cde7421b9d6c96ddab4468e3e05b899778eaf